targets/TARGET_Silicon_Labs/TARGET_EFM32/trng/sl_trng.c

```diff
@@ -0,0 +1,274 @@
+    device->KEY1 = *_key++;
+    device->KEY2 = *_key++;
+    device->KEY3 = *_key++;
+
+    return 0;
+}
+
+static int sl_trng_check_status( TRNG_TypeDef *device )
+{
+    uint32_t status = device->STATUS;
+
+    if ( (status & (TRNG_STATUS_PREIF
+                    | TRNG_STATUS_REPCOUNTIF
+                    | TRNG_STATUS_APT64IF
+                    | TRNG_STATUS_APT4096IF
+                    | TRNG_STATUS_ALMIF)) == 0 )
+    {
+        /* No errors */
+        return 0;
+    }
+
+    if ( status & TRNG_STATUS_PREIF )
+    {
+        /* On a preliminary noise alarm we clear the FIFO and clear
+         * the alarm. The preliminary noise alarm is not critical. */
+        status &= ~TRNG_STATUS_PREIF;
+        device->STATUS = status;
+        sl_trng_clear_fifo(device);
+        return SL_TRNG_ERR_PRELIMINARY_NOISE_ALARM;
+    }
+    else
+    {
+        /* Clear alarm conditions by doing a TRNG soft reset. */
+        sl_trng_soft_reset( device );
+        if ( status & TRNG_STATUS_REPCOUNTIF )
+        {
+            return SL_TRNG_ERR_REPETITION_COUNT_TEST_FAILED;
+        }
+        if ( status & TRNG_STATUS_APT64IF )
+        {
+            return SL_TRNG_ERR_ADAPTIVE_PROPORTION_TEST_64_FAILED;
+        }
+        if ( status & TRNG_STATUS_APT4096IF )
+        {
+            return SL_TRNG_ERR_ADAPTIVE_PROPORTION_TEST_4096_FAILED;
+        }
+        if ( status & TRNG_STATUS_ALMIF )
+        {
+            return SL_TRNG_ERR_NOISE_ALARM;
+        }
+    }
+
+    return 0;
+}
+
+static void sl_trng_read_chunk( TRNG_TypeDef *device,
+                                unsigned char *output,
+                                size_t len )
+{
+    uint32_t * out32 = (uint32_t *) output;
+    uint32_t tmp;
+
+    /* Read known good available data. */
+    while ( len >= 4)
+    {
+        *out32++ = device->FIFO;
+        len -= 4;
+    }
+
+    /* Handle the case where len is not a multiple of 4. */
+    if ( len < 4 )
+    {
+        tmp = device->FIFO;
+        memcpy((uint8_t *)out32, (const uint8_t *) &tmp, len);
+    }
+}
+
+int sl_trng_poll( TRNG_TypeDef *device,
+                  unsigned char *output,
+                  size_t len,
+                  size_t *olen )
+{
+    size_t output_len = 0;
+    size_t chunk_len = 0;
+    size_t available;
+    int ret = 0;
+
+    while (output_len < len)
+    {
+        available = device->FIFOLEVEL * 4;
+        if (available == 0)
+        {
+            break;
+        }
+
+#if !defined(SL_TRNG_IGNORE_ALL_ALARMS)
+        /* Check status for current data in FIFO
+         * and handle any error conditions. */
+        ret = sl_trng_check_status( device );
+#if defined(SL_TRNG_IGNORE_NOISE_ALARMS)
+        /* Ignore noise alarms by returning 0 (OK) if they occur and
+         * keeping the already generated random data. */
+        if ( (ret == SL_TRNG_ERR_PRELIMINARY_NOISE_ALARM) ||
+             (ret == SL_TRNG_ERR_NOISE_ALARM) )
+        {
+            ret = 0;
+            continue;
```

Comment:
Like [yanesca](https://github.com/ARMmbed/mbed-os/pull/4051/files#r118708273) I wonder why there's a compilation option to ignore alarms. Assuming that the mode `SL_TRNG_IGNORE_NOISE_ALARMS` is useful, I don't understand the logic here. If there is a physical condition that causes the alarm to keep being raised, wouldn't this cause a tight loop?